[pytorch-rpc] WireSerializer should check has_storage()

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34626 [pytorch-rpc] WireSerializer should check has_storage()**

We need to check has_storage() before looking at it in
cloneSparseTensors(), to avoid gratuitously throwing.

Add some basic test coverage (though some work work until
Pickler supports sparse tensors)

Differential Revision: [D20399971](https://our.internmc.facebook.com/intern/diff/D20399971/)

Commit:
[pytorch-rpc] WireSerializer should check has_storage()

We need to check has_storage() before looking at it in
cloneSparseTensors(), to avoid gratuitously throwing.

Ideally, we'd add a test for this (I wrote one up but had to disable it),
but won't work until JIT Pickler supports sparse tensors.

Differential Revision: [D20399971](https://our.internmc.facebook.com/intern/diff/D20399971/)

[ghstack-poisoned]

Commit:
Update on "[pytorch-rpc] WireSerializer should check has_storage()"


We need to check has_storage() before looking at it in
cloneSparseTensors(), to avoid gratuitously throwing.

Ideally, we'd add a test for this (I wrote one up but had to disable it),
but won't work until JIT Pickler supports sparse tensors.

Differential Revision: [D20399971](https://our.internmc.facebook.com/intern/diff/D20399971/)

[ghstack-poisoned]

pritamdamania87 commented on test/cpp/rpc/test_wire_serialization.cpp:
```diff
@@ -49,3 +49,12 @@ TEST(WireSerialize, RecopySparseTensors) {
+  auto ser = torch::distributed::rpc::wireSerialize({}, {main.to(at::kSparse)});
+  auto deser = torch::distributed::rpc::wireDeserialize(ser.data(), ser.size());
+  EXPECT_TRUE(torch::equal(main, deser.second[0]));
+}
```
Can we add a unit test that directly tests `cloneSparseTensors` for now and later replace it with this test?

jjlilley replied:
Good point, I forgot that we were exporting that symbol.

Commit:
Update on "[pytorch-rpc] WireSerializer should check has_storage()"


We need to check has_storage() before looking at it in
cloneSparseTensors(), to avoid gratuitously throwing.

Ideally, we'd add a test for this (I wrote one up but had to disable it),
but won't work until JIT Pickler supports sparse tensors.

Differential Revision: [D20399971](https://our.internmc.facebook.com/intern/diff/D20399971/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -220,6 +220,9 @@ c10::List<at::Tensor> cloneSparseTensors(
   // ~1% of the underlying Storage.
   auto worthRecopying = [](const at::Tensor& t) -> bool {
+    if (!t.has_storage()) {
+      return false; // avoid throwing below.
```
BTW, as JIT does not support sparse tensors yet, what error will user see after this change?

jjlilley replied:
When I ran this last night, they see the error from JIT Pickler, rather than this line, where we were looking at the storage without it being there.

mrshenli reviewed (commented):
Shall we add a Python test for this? We can use similar code from #33939

mrshenli reviewed (approved):
Oh, if the change if localized to wireSerializer, the current version LGTM. We can add Python test and error message after this PR.

jjlilley commented:
Thanks for reviewing!
